server/parsec/components/shamir.py

```diff
@@ -0,0 +1,130 @@
+                client_ctx.organization_id,
+                client_ctx.user_id,
+                client_ctx.device_id,
+                client_ctx.device_verify_key,
+                req.setup,
+            ):
+                case None:
+                    return authenticated_cmds.latest.shamir_recovery_setup.RepOk()
+                case (
+                    ShamirAddRecoverySetupBadOutcome.INVALID_DATA
+                    | ShamirAddRecoverySetupBadOutcome.AUTHOR_NOT_FOUND
+                    | ShamirAddRecoverySetupBadOutcome.ORGANIZATION_EXPIRED
+                    | ShamirAddRecoverySetupBadOutcome.AUTHOR_REVOKED
+                ):
+                    return authenticated_cmds.latest.shamir_recovery_setup.RepInvalidData()
+                case ShamirAddRecoverySetupBadOutcome.ALREADY_SET:
+                    return authenticated_cmds.latest.shamir_recovery_setup.RepShamirSetupAlreadyExists()
+                case ShamirAddRecoverySetupBadOutcome.INVALID_CERTIFICATE:
+                    return authenticated_cmds.latest.shamir_recovery_setup.RepInvalidCertificate()
+                case _:  # if this happens, it means that an error case is not covered above
+                    raise NotImplementedError
+
+    # async def test_dump_current_shamir(
+    #     self, organization_id: OrganizationID
+    # ) -> dict[UserID, ShamirDump]:
+    #     raise NotImplementedError
+
+    async def remove_recovery_setup(
+        self,
+        organization_id: OrganizationID,
+        author: UserID,
+    ) -> None:
+        raise NotImplementedError
+
+    async def add_recovery_setup(
+        self,
+        organization_id: OrganizationID,
+        author: UserID,
+        device: DeviceID,
+        author_verify_key: VerifyKey,
+        setup: authenticated_cmds.latest.shamir_recovery_setup.ShamirRecoverySetup,
+    ) -> None | ShamirAddRecoverySetupBadOutcome:
+        raise NotImplementedError
+
+
+class ShamirAddRecoverySetupBadOutcome(BadOutcomeEnum):
+    INVALID_DATA = auto()
+    ORGANIZATION_NOT_FOUND = auto()
+    ORGANIZATION_EXPIRED = auto()
+    AUTHOR_NOT_FOUND = auto()
+    AUTHOR_REVOKED = auto()
+    ALREADY_SET = auto()
+    INVALID_CERTIFICATE = auto()
+    AUTHOR_FROZEN = auto()
+    INVALID_RECIPIENT = auto()
+
+
+class VerifyCertificatesBadOutcome(BadOutcomeEnum):
+    INVALID_DATA = auto()
+
+
+# Check internal consistency of certificate
+def verify_certificates(
+    setup: authenticated_cmds.latest.shamir_recovery_setup.ShamirRecoverySetup,
+    author: DeviceID,
+    author_verify_key: VerifyKey,
+) -> tuple[ShamirRecoveryBriefCertificate, dict[UserID, bytes]] | VerifyCertificatesBadOutcome:
+    share_certificates: dict[UserID, bytes] = {}
+    try:
+        brief_certificate = ShamirRecoveryBriefCertificate.verify_and_load(
+            setup.brief, author_verify_key, expected_author=author
+        )
+    except ValueError:
+        return VerifyCertificatesBadOutcome.INVALID_DATA
```

Comment:
The other methods try to stick to a consistent naming. Here's what certificate verification looks like for `realm_create`:
```python
class RealmCreateValidateBadOutcome(BadOutcomeEnum):
    INVALID_CERTIFICATE = auto()
    TIMESTAMP_MISMATCH = auto()
    INVALID_ROLE = auto()
    USER_ID_MISMATCH = auto()


def realm_create_validate(
    now: DateTime,
    expected_author: DeviceID,
    author_verify_key: VerifyKey,
    realm_role_certificate: bytes,
) -> RealmRoleCertificate | TimestampOutOfBallpark | RealmCreateValidateBadOutcome:
```

That means the following changes:
- rename `VerifyCertificatesBadOutcome` to `ShamirAddRecoverySetupValidateBadOutcome`
- rename `verify_certificates` to `shamir_add_recovery_setup_validate`
- add and manage `TimestampOutOfBallpark` inside the validate method